ilastik/applets/dataSelection/opDataSelection.py

```diff
@@ -19,195 +19,353 @@
 #          http://ilastik.org/license.html
 ###############################################################################
 import glob
-import numpy
 import os
 import uuid
-import vigra
 import copy
+from enum import Enum, unique
+from typing import List, Iterable, Tuple, Callable
+from numbers import Number
+import re
+import functools
+from pathlib import Path
+import errno
+
+import numpy
+import vigra
+import h5py
+import z5py
 
 from lazyflow.graph import Operator, InputSlot, OutputSlot, OperatorWrapper
 from lazyflow.metaDict import MetaDict
-from lazyflow.operators.ioOperators import (
-    OpStreamingH5N5Reader, OpStreamingH5N5SequenceReaderS, OpInputDataReader
-)
+from lazyflow.operators.ioOperators import OpStreamingH5N5Reader
+from lazyflow.operators.ioOperators import OpInputDataReader, OpH5N5WriterBigDataset
 from lazyflow.operators.valueProviders import OpMetadataInjector, OpZeroDefault
 from lazyflow.operators.opArrayPiper import OpArrayPiper
 from ilastik.applets.base.applet import DatasetConstraintError
 
 from ilastik.utility import OpMultiLaneWrapper
+from lazyflow.graph import Graph
 from lazyflow.utility import PathComponents, isUrl, make_absolute
+from lazyflow.utility.pathHelpers import splitPath, globH5N5, globNpz
 from lazyflow.utility.helpers import get_default_axisordering
 from lazyflow.operators.opReorderAxes import OpReorderAxes
+from lazyflow.graph import Graph
 
+def getTypeRange(numpy_type):
+    type_info = numpy.iinfo(numpy_type)
+    return (type_info.min, type_info.max)
 
 class DatasetInfo(object):
     """
     Struct-like class for describing dataset info.
     """
-    class Location(object):
-        FileSystem = 0
+    @unique
+    class Location(Enum):
+        FileSystem = 0 #deprecated
         ProjectInternal = 1
         PreloadedArray = 2
-
-    def __init__(self, filepath=None, jsonNamespace=None, cwd=None,
-                 preloaded_array=None, sequence_axis=None):
+        FileSystemRelativePath = 3
+        FileSystemAbsolutePath = 4
+
+    def __init__(
+        self,
+        *,
+        filepath:str=None,
+        project_file:h5py.File=None,
+        preloaded_array=None,
+        sequence_axis=None,
+        allowLabels=True,
+        subvolume_roi=None,
+        location=None,
+        axistags=None,
+        fill_in_dummy_axes:bool=False,
+        display_mode='default',
+        nickname='',
+        original_axistags=None,
+        normalizeDisplay:bool=None,
+        drange:Tuple[Number, Number]=None,
+        jsonNamespace=None,
+    ):
```

Comment:
I wonder if we could create a number of factory functions for `DatasetInfo` with different parameters, based on `location` cases (?). 

In any case, this constructor takes a lot of parameters that influence control flow in non-trivial ways. Is it possible to somehow simplify this?

Reply:
I guess I was too tied to the legacy implementation... But you;re right. I;ve split `DatasetInfo` into a few subclasses that do their own stuff and that take their own, specific arguements, rather than just having all of them available in the generic `DatasetInfo`